### PR TITLE
Fix for binary data representation when sending data through HTTP

### PR DESCRIPTION
**Pull request by MayankSingh-NEC**
This PR fixes https://github.com/telefonicaid/iotagent-json/issues/690

### CODE REVIEW

**Comment by fgalan** [comment]
It seems unit tests are not working. Could you have a look, please?

**Comment by MayankSingh-NEC** [comment]
> It seems unit tests are not working. Could you have a look, please?

Hi @fgalan , I've checked and found that the unit tests are not failing due to this PR.

I've performed the unit tests without applying my fix and the same UT are failing for the master branch, 1.25.0, and 1.24.0 as well.

**Comment by fgalan** [comment]
That's weird... note that GitHub confirms that at master branch all the tests are working at the present moment (i.e. last commit in master):

imagen

(The red X is due to fails in the coverage job, but tests are green tick)

So maybe something in your environment is making tests at master to fail. But the fail in your PR branch is being confirmed by GitHub.

**Comment by fgalan** [comment]
I'll re-run both tests at master and in your branch to be sure. I'll tell you in some minutes...

**Comment by fgalan** [comment]
> I'll re-run both tests at master and in your branch to be sure. I'll tell you in some minutes...

* At your branch, same result **NOK**: https://github.com/telefonicaid/iotagent-json/actions/runs/4004935581
* At master, same result **OK**: https://github.com/telefonicaid/iotagent-json/actions/runs/3998343745

**Comment by MayankSingh-NEC** [comment]
Hi @fgalan , yes, I've tested and verified again that these unit test cases are now working fine. Also, I'm looking into my fix and the test cases that are failing.

Sorry for inconvenience.

**Comment by fgalan** [comment]
It seems tests are still failing after last commits.

**Comment by MayankSingh-NEC** [comment]
Hi @fgalan , I have the PR. Please approve the workflow.
Thank you.

**Comment by fgalan** [comment]
It seems tests are now passing. Great!

We will have a look to the PR when we have some time. Please hold on :)

File: lib/commonBindings.js
Comment (mapedraza) [inline lib/commonBindings.js:59]: Seems there are leftovers included in the code

File: test/unit/ngsiv2/contextRequests/singleMeasuresRawTypes.json
Comment (mapedraza) [inline test/unit/ngsiv2/contextRequests/singleMeasuresRawTypes.json:4]: This is not right, the expected result for the request is "3332", which is an string with the hex representation of the data sent ( a string, "32")

Comment (mapedraza) [reply test/unit/ngsiv2/contextRequests/singleMeasuresRawTypes.json:4]: Here the part of the test that send the request:
https://github.com/telefonicaid/iotagent-json/blob/36bcc64fee817c4289f048b9ded3b2ea6fe5f4d9/test/unit/ngsiv2/HTTP_receive_measures-test3.js#L190-L205

**Comment by mapedraza** [comment]
The expected value is a string containing the hex representation of the data received

**Comment by MayankSingh-NEC** [comment]
Hi @mapedraza , I've made some changes according to the expected value as per your suggestion.

Please have a look and do let me know if any other changes required. Thanks :))

File: test/unit/ngsiv2/contextRequests/singleMeasuresRawTypes.json
Comment (mapedraza) [inline test/unit/ngsiv2/contextRequests/singleMeasuresRawTypes.json:6]: Analysing this again, the test that is expecting this answer is:
https://github.com/telefonicaid/iotagent-json/blob/36bcc64fee817c4289f048b9ded3b2ea6fe5f4d9/test/unit/ngsiv2/HTTP_receive_measures-test3.js#L190-L205

Character 3 and 2 are `0x33` and `0x32` in hex representation.

Since the expected result is:
```
     "value": "3332"
````

It seems is the expected result.

**Comment by mapedraza** [comment]
[CHANGES_NEXT_RELEASE](https://github.com/telefonicaid/iotagent-json/blob/master/CHANGES_NEXT_RELEASE) entry is mising

**Review by mapedraza — APPROVED** [review]
LGTM (pending [CHANGES_NEXT_RELEASE](https://github.com/telefonicaid/iotagent-json/blob/master/CHANGES_NEXT_RELEASE) entry)

**Comment by MayankSingh-NEC** [comment]
> LGTM (pending [CHANGES_NEXT_RELEASE](https://github.com/telefonicaid/iotagent-json/blob/master/CHANGES_NEXT_RELEASE) entry)

Hi @mapedraza , Thanks for reviewing this PR.
I've completed the pending [CHANGES_NEXT_RELEASE](https://github.com/telefonicaid/iotagent-json/blob/master/CHANGES_NEXT_RELEASE) entry task.

**Comment by mapedraza** [comment]
@MayankSingh-NEC a twin PR fixing same issue on https://github.com/telefonicaid/iotagent-ul should be raised. This would keep aligned IoTAS behaviour.

**Comment by MayankSingh-NEC** [comment]
> @MayankSingh-NEC a twin PR fixing same issue on https://github.com/telefonicaid/iotagent-ul should be raised. This would keep aligned IoTAS behaviour.

Hi @mapedraza , Currently, I'm working on the same for UL. Please let me know if anything left to do in this PR. Meanwhile, please merge this PR, if it looks good. Thanks !! :))

**Comment by fgalan** [comment]
> > @MayankSingh-NEC a twin PR fixing same issue on https://github.com/telefonicaid/iotagent-ul should be raised. This would keep aligned IoTAS behaviour.
> 
> Hi @mapedraza , Currently, I'm working on the same for UL. Please let me know if anything left to do in this PR. Meanwhile, please merge this PR, if it looks good. Thanks !! :))

The plan is as follows:

1. Create and review the PR for IOTA-UL
2. Once both the PR in IOTA-UL and IOTA-JSON gets LGTM status, both will be merged at the same time

If we merge this PR before having the IOTA-UL PR at the same state, the IOTAs would be unbalanced. Although preventively this PR could get LGTM, sometimes new things arise in the twin PR so it is wise to keep both open until the end and merge at the same time both.

File: lib/bindings/HTTPBinding.js
Comment (mapedraza) [inline lib/bindings/HTTPBinding.js:297]: After reviewing this PR, the hex transformation was moved to HTTPBinding file. This is not correct, since it has to be valid also for MQTT messages. 

All changes should be included in `lib/commonBindings.js`file

**Review by mapedraza — CHANGES_REQUESTED** [review]

**Comment by MayankSingh-NEC** [comment]
Hi @mapedraza , I've committed the change. 

Yes, earlier it was only fixed for HTTP and not for the MQTT.
Now, it's working fine for both the HTTP & MQTT.

**Comment by MayankSingh-NEC** [comment]
> @MayankSingh-NEC test are failling, please, could you check them? It seems somethings is failling. Additionally, changes still present in file lib/bindings/HTTPBinding.js

Hi @mapedraza , I have checked and verified that these UTs aren't failing in my system with my code fix. All UTs are getting passed. This might be failing for some other reasons. Please find below attachments:


According to my understanding, The fix given in HTTP bindings file was to convert binary data coming from the HTTP. Similarly, in a commonBindings file, convert to binary data coming from MQTT.

**Comment by mapedraza** [comment]
Could you try to rebase or merge master branch here? #719 Fix bidirectionality tests, maybe this should solve your issues

File: CHANGES_NEXT_RELEASE
Comment (fgalan) [inline CHANGES_NEXT_RELEASE:5]: This content seems to be old. The content of this file should be:

* Existing content in CHANGES_NEXT_RELEASE at this moment (can be check at https://github.com/telefonicaid/iotagent-json/blob/master/CHANGES_NEXT_RELEASE)
* The entries corresponding to the fixes in this PR

Comment (MayankSingh-NEC) [reply CHANGES_NEXT_RELEASE]: I have updated in the CHANGES_NEXT_RELEASE.

File: CHANGES_NEXT_RELEASE
Comment (fgalan) [inline CHANGES_NEXT_RELEASE:2]:
```suggestion
- Fix: for the binary data representation when sending data through MQTT and HTTP (#690)

```

File: lib/bindings/HTTPBinding.js
Comment (AlvaroVega) [reply lib/bindings/HTTPBinding.js:132]: But it seems that  `data` is not stored in req,jsonPayload when new iff and is just print by a log and any more.

**Comment by MayankSingh-NEC** [comment]
Hi @mapedraza and @fgalan ,

My apology regarding this closed PR as I was fixing but due to some technical glitch it is closed. Also, other PRs are effected by it. I will raise a separate PR for the same. Thank you :))

**Comment by fgalan** [comment]
> I will raise a separate PR for the same.

Please include a reference to PR #696 in the new one you are going to open, for traceability. Thanks!

**Pull request closed without merging**
